File: core/runtime/examples/embed.c
#include "hoplite_runtime.h"

#include <assert.h>
#include <stddef.h>
#include <stdint.h>
#include <string.h>

static hoplite_slice_t text(const char *value) {
  hoplite_slice_t slice = {
      .data = (const uint8_t *)value,
      .len = strlen(value),
  };
  return slice;
}

static int32_t raw_field(void *context, uint32_t field,
                         hoplite_slice_t *value) {
  if (context == NULL || value == NULL) {
    return HOPLITE_RAW_FIELD_ERROR;
  }
  value->data = NULL;
  value->len = 0;
  if (field != HOPLITE_RAW_FIELD_SCHEME) {
    return HOPLITE_RAW_FIELD_UNAVAILABLE;
  }
  *value = text((const char *)context);
  return HOPLITE_RAW_FIELD_OK;
}

int main(void) {
  static const char source[] =
      "(ns embedding.example (:require [std.foundation :refer :all])) "
      "(defn respond [exchange] "
      "{:status 200 :headers {\"content-type\" \"text/plain\"} "
      ":body (or (:scheme exchange) \"missing\")})";
  static const char function[] = "embedding.example/respond";
  static const char scheme[] = "https";

  assert(hoplite_abi_version() >= 5);

  hoplite_runtime_t *runtime = hoplite_runtime_new();
  assert(runtime != NULL);
  assert(hoplite_bootstrap_modules(runtime, (const uint8_t *)source,
                                   strlen(source)) == 0);

  uint64_t handler = hoplite_handler_prepare(
      runtime, (const uint8_t *)function, strlen(function));
  assert(handler != 0);

  hoplite_raw_request_v1_t raw = {
      .context = (void *)scheme,
      .field = raw_field,
  };
  hoplite_request_v4_t request = {
      .request =
          {
              .request =
                  {
                      .context = NULL,
                      .method = text("GET"),
                      .uri = text("/embed"),
                      .path = text("/embed"),
                      .query_string = text(""),
                      .remote_address = text("127.0.0.1"),
                      .header_count = 0,
                      .header_at = NULL,
                  },
              .body = NULL,
              .max_body_bytes = 0,
              .max_chunk_bytes = 0,
              .require_declared_length = 0,
          },
      .raw = &raw,
  };
  hoplite_outcome_v2_t outcome = {0};
  assert(hoplite_handler_invoke_v4(runtime, handler, 0, &request,
                                   &outcome) == 0);
  assert(outcome.kind == 1);
  assert(outcome.id != 0);

  uint16_t status = 0;
  assert(hoplite_response_status_v2(runtime, outcome.id, &status) == 0);
  assert(status == 200);

  hoplite_slice_t body = {0};
  assert(hoplite_response_body_v2(runtime, outcome.id, &body) == 0);
  assert(body.data != NULL);
  assert(body.len == strlen(scheme));
  assert(memcmp(body.data, scheme, body.len) == 0);

  assert(hoplite_response_close_v2(runtime, outcome.id) == 0);
  assert(hoplite_handler_close(runtime, handler) == 0);
  hoplite_runtime_free(runtime);
  return 0;
}

File: core/runtime/examples/embed.rs
use hoplite_runtime::{
    hoplite_abi_version, hoplite_bootstrap_modules, hoplite_handler_close,
    hoplite_handler_invoke_v2, hoplite_handler_prepare, hoplite_response_body_v2,
    hoplite_response_close_v2, hoplite_response_status_v2, hoplite_runtime_free,
    hoplite_runtime_new, HopliteOutcomeV2, HopliteRequestV2, HopliteSlice,
};
use std::ffi::c_void;
use std::ptr;
use std::slice;

fn text(value: &str) -> HopliteSlice {
    HopliteSlice {
        data: value.as_ptr(),
        len: value.len(),
    }
}

fn main() {
    assert!(hoplite_abi_version() >= 5);

    let runtime = hoplite_runtime_new();
    assert!(!runtime.is_null());

    unsafe {
        let source = r#"
(ns embedding.example (:require [std.foundation :refer :all]))
(defn respond [request]
  {:status 200
   :headers {"content-type" "text/plain"}
   :body "embedded"})
"#;
        assert_eq!(
            hoplite_bootstrap_modules(runtime, source.as_ptr(), source.len()),
            0
        );

        let function = "embedding.example/respond";
        let handler = hoplite_handler_prepare(runtime, function.as_ptr(), function.len());
        assert_ne!(handler, 0);

        let request = HopliteRequestV2 {
            context: ptr::null_mut::<c_void>(),
            method: text("GET"),
            uri: text("/embed"),
            path: text("/embed"),
            query_string: text(""),
            remote_address: text("127.0.0.1"),
            header_count: 0,
            header_at: None,
        };
        let mut outcome = HopliteOutcomeV2 { kind: 0, id: 0 };
        assert_eq!(
            hoplite_handler_invoke_v2(runtime, handler, 1, &request, &mut outcome),
            0
        );
        assert_eq!(outcome.kind, 1);
        assert_ne!(outcome.id, 0);

        let mut status = 0_u16;
        assert_eq!(
            hoplite_response_status_v2(runtime, outcome.id, &mut status),
            0
        );
        assert_eq!(status, 200);

        let mut body = HopliteSlice {
            data: ptr::null(),
            len: 0,
        };
        assert_eq!(hoplite_response_body_v2(runtime, outcome.id, &mut body), 0);
        assert!(!body.data.is_null());
        assert_eq!(slice::from_raw_parts(body.data, body.len), b"embedded");

        assert_eq!(hoplite_response_close_v2(runtime, outcome.id), 0);
        assert_eq!(hoplite_handler_close(runtime, handler), 0);
        hoplite_runtime_free(runtime);
    }
}
